modules/effects/arrow-effects-data/src/main/kotlin/arrow/effects/typeclasses/Concurrent.kt

```diff
@@ -722,6 +724,87 @@ interface Concurrent<F> : Async<F> {
 
   override fun <B> binding(c: suspend MonadContinuation<F, *>.() -> B): Kind<F, B> =
     bindingCancellable { c() }.a
+
+  /**
+   *  Sleeps for a given [duration] without blocking a thread.
+   *  Used to derive combinators [timeout], [timeoutTo] and can be used to created timed events.
+   *
+   * ```kotlin:ank:playground
+   * import arrow.effects.*
+   * import arrow.effects.extensions.io.concurrent.concurrent
+   *
+   * fun main(args: Array<String>) {
+   *   //sampleStart
+   *   fun <F> Concurrent<F>.delayHelloWorld(): Kind<F, Unit> =
+   *     sleep(3.seconds).flatMap {
+   *       delay { println("Hello World!") }
+   *     }
+   *   //sampleEnd
+   *   IO.unsafeRunBlocking(IO.concurrent().delayHelloWorld())
+   * }
+   * ```
+   * @see timeout
+   * @see timeoutTo
+   **/
+  fun sleep(duration: Duration): Kind<F, Unit> = ConcurrentSleep(duration)
+
+  /**
+   * Returns the result of [this] within the specified [duration] or the [default] value.
+   *
+   * ```kotlin:ank:playground
+   * import arrow.effects.*
+   * import arrow.effects.extensions.io.concurrent.concurrent
+   * import arrow.effects.typeclasses.seconds
+   *
+   * fun main(args: Array<String>) {
+   *   //sampleStart
+   *   fun <F> Concurrent<F>.timedOutWorld(): Kind<F, Unit> {
+   *     val world = sleep(3.seconds).flatMap { delay { println("Hello World!") } }
+   *     val fallbackWorld = delay { println("Hello from the backup") }
+   *     return world.timeoutTo(fallbackWorld, 1.seconds)
+   *   }
+   *   //sampleEnd
+   *   IO.unsafeRunBlocking(IO.concurrent().timedOutWorld())
+   * }
+   * ```
+   * @see timeout
+   **/
+  fun <A> Kind<F, A>.timeoutTo(default: Kind<F, A>, duration: Duration): Kind<F, A> =
```

Comment:
Is this a better name?
```kotlin
fun <A> Kind<F, A>.waitFor(duration: Duration, otherwise: Kind<F, A>): Kind<F, A>

IO { expensive() }.waitFor(4.seconds, otherwise = IO { cached() })
```

Reply:
What about replacing both `timeout` and `timeoutTo` with the following signature?

```
fun <A> Kind<F, A>.waitFor(duration: Duration, otherwise: Kind<F, A> = just(TimeoutException)): Kind<F, A>
```

Reply:
looks great

Reply:
Not a fan of defaults in interfaces. They cannot be overriden.